From 8b1fd0dd72f09fc5a7395a96636f64ffad953ff9 Mon Sep 17 00:00:00 2001
From: "Tu .V Nguyen [FGA.AIS]" <TuNV16@fsoft.com.vn>
Date: Thu, 11 Oct 2018 10:43:25 +0700
Subject: [PATCH 406/628] watchdog: renesas-wdt: fix reboot issue on G1M-N

	Revert 5 functions from CIP25 below:
		1. static int rwdt_init_timeout(struct watchdog_device *wdev)
		2. static int rwdt_ping(struct watchdog_device *wdev)
		3. static int rwdt_start(struct watchdog_device *wdev)
		4. static int rwdt_stop(struct watchdog_device *wdev)
		5. static int rwdt_probe(struct platform_device *pdev)

Signed-off-by: Tu .V Nguyen [FGA.AIS] <TuNV16@fsoft.com.vn>
---
 drivers/watchdog/renesas_wdt.c | 52 ++++++++++++++++--------------------------
 1 file changed, 20 insertions(+), 32 deletions(-)

diff --git a/drivers/watchdog/renesas_wdt.c b/drivers/watchdog/renesas_wdt.c
index c086afe..f32bdff 100644
--- a/drivers/watchdog/renesas_wdt.c
+++ b/drivers/watchdog/renesas_wdt.c
@@ -58,6 +58,8 @@ struct rwdt_priv {
 	u16 time_left;
 	u8 cks;
 	struct notifier_block restart_handler;
+	struct clk *clk;
+	unsigned int clks_per_sec;
 };
 
 static void rwdt_write(struct rwdt_priv *priv, u32 val, unsigned int reg)
@@ -70,30 +72,28 @@ static void rwdt_write(struct rwdt_priv *priv, u32 val, unsigned int reg)
 	writel_relaxed(val, priv->base + reg);
 }
 
-static int rwdt_set_timeout(struct watchdog_device *wdev, unsigned int seconds)
+static int rwdt_init_timeout(struct watchdog_device *wdev)
 {
 	struct rwdt_priv *priv = watchdog_get_drvdata(wdev);
 
-	wdev->timeout = seconds;
-	rwdt_write(priv, 65536 - MUL_BY_CLKS_PER_SEC(priv, wdev->timeout), RWTCNT);
+	rwdt_write(priv, 65536 - wdev->timeout * priv->clks_per_sec, RWTCNT);
 
 	return 0;
 }
 
 static int rwdt_ping(struct watchdog_device *wdev)
 {
-	return rwdt_set_timeout(wdev, wdev->timeout);
+	return rwdt_init_timeout(wdev);
 }
 
 static int rwdt_start(struct watchdog_device *wdev)
 {
 	struct rwdt_priv *priv = watchdog_get_drvdata(wdev);
 
-	pm_runtime_get_sync(wdev->parent);
+	clk_prepare_enable(priv->clk);
 
-	rwdt_write(priv, 0, RWTCSRB);
 	rwdt_write(priv, priv->cks, RWTCSRA);
-	rwdt_set_timeout(wdev, wdev->timeout);
+	rwdt_init_timeout(wdev);
 
 	while (readb_relaxed(priv->base + RWTCSRA) & RWTCSRA_WRFLG)
 		cpu_relax();
@@ -108,7 +108,7 @@ static int rwdt_stop(struct watchdog_device *wdev)
 	struct rwdt_priv *priv = watchdog_get_drvdata(wdev);
 
 	rwdt_write(priv, priv->cks, RWTCSRA);
-	pm_runtime_put(wdev->parent);
+	clk_disable_unprepare(priv->clk);
 
 	return 0;
 }
@@ -193,14 +193,13 @@ static const struct watchdog_ops rwdt_ops = {
 	.stop = rwdt_stop,
 	.ping = rwdt_ping,
 	.get_timeleft = rwdt_get_timeleft,
-	.set_timeout = rwdt_set_timeout,
+	.set_timeout = rwdt_init_timeout,
 };
 
 static int rwdt_probe(struct platform_device *pdev)
 {
 	struct rwdt_priv *priv;
 	struct resource *res;
-	struct clk *clk;
 	unsigned long clks_per_sec;
 	int ret, i;
 
@@ -213,36 +212,34 @@ static int rwdt_probe(struct platform_device *pdev)
 	if (IS_ERR(priv->base))
 		return PTR_ERR(priv->base);
 
-	clk = devm_clk_get(&pdev->dev, NULL);
-	if (IS_ERR(clk))
-		return PTR_ERR(clk);
-
-	pm_runtime_enable(&pdev->dev);
-	pm_runtime_get_sync(&pdev->dev);
-	priv->clk_rate = clk_get_rate(clk);
-	priv->wdev.bootstatus = (readb_relaxed(priv->base + RWTCSRA) &
-				RWTCSRA_WOVF) ? WDIOF_CARDRESET : 0;
-	pm_runtime_put(&pdev->dev);
+	priv->clk = devm_clk_get(&pdev->dev, NULL);
+	if (IS_ERR(priv->clk))
+		return PTR_ERR(priv->clk);
 
+	priv->clk_rate = clk_get_rate(priv->clk);
 	if (!priv->clk_rate) {
 		ret = -ENOENT;
 		goto out_pm_disable;
 	}
 
 	for (i = ARRAY_SIZE(clk_divs) - 1; i >= 0; i--) {
-		clks_per_sec = priv->clk_rate / clk_divs[i];
-		if (clks_per_sec && clks_per_sec < 65536) {
+		clks_per_sec = DIV_ROUND_UP(priv->clk_rate, clk_divs[i]);
+		if (clks_per_sec) {
+			priv->clks_per_sec = clks_per_sec;
 			priv->cks = i;
 			break;
 		}
 	}
 
-	if (i < 0) {
+	if (!clks_per_sec) {
 		dev_err(&pdev->dev, "Can't find suitable clock divider\n");
 		ret = -ERANGE;
 		goto out_pm_disable;
 	}
 
+	pm_runtime_enable(&pdev->dev);
+	pm_runtime_get_sync(&pdev->dev);
+
 	priv->wdev.info = &rwdt_ident,
 	priv->wdev.ops = &rwdt_ops,
 	priv->wdev.parent = &pdev->dev;
@@ -263,15 +260,6 @@ static int rwdt_probe(struct platform_device *pdev)
 	if (ret < 0)
 		goto out_pm_disable;
 
-	priv->restart_handler.notifier_call = rwdt_restart;
-	priv->restart_handler.priority = 0;
-	ret = register_restart_handler(&priv->restart_handler);
-	if (ret) {
-		dev_err(&pdev->dev, "can't register restart handler (err=%d)\n",
-			ret);
-		priv->restart_handler.notifier_call = NULL;
-	}
-
 	/*
 	 * register restart handler base on machine here
 	 * same ARM core architecture (e.g ARM cortex A15) can use same handler
-- 
2.7.4

